Improve the error message note when an Error Union Type is found but the compiler expected a normal type

### Zig Version

0.12.0-dev.2341+92211135

### Steps to Reproduce and Observed Output

```zig
const std = @import("std");
const debug = std.debug;
const heap = std.heap;

pub fn main() !void {
    const allocator = heap.page_allocator;
    const list = allocator.alloc(u8, 10);

    for (list) |item| {
        debug.print("{s}\n", .{item});
    }
}
```

```console
$ zig run test.zig
test.zig:10:10: error: type 'error{OutOfMemory}![]u8' is not indexable and not a range
    for (list) |item| {
         ^~~~
test.zig:10:10: note: for loop operand must be a range, array, slice, tuple, or vector
```
```


### Expected Output

The compiler message should add an additional `note`, explaining that the user should use `try list`.

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -4221,6 +4221,11 @@ fn zirForLen(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.
                     const msg = try sema.errMsg(block, arg_src, "type '{}' is not indexable and not a range", .{object_ty.fmt(sema.mod)});
                     errdefer msg.destroy(sema.gpa);
                     try sema.errNote(block, arg_src, msg, "for loop operand must be a range, array, slice, tuple, or vector", .{});
+
+                    if (object_ty.zigTypeTag(mod) == .ErrorUnion) {
+                        try sema.errNote(block, arg_src, msg, "consider using 'try', 'catch', or 'if'", .{});
+                    }
+
                     break :msg msg;
                 };
                 return sema.failWithOwnedErrorMsg(block, msg);

--- a/test/cases/compile_errors/for_loop_error_union.zig
+++ b/test/cases/compile_errors/for_loop_error_union.zig
@@ -10,6 +10,6 @@ export fn a() void {
 // backend=stage2
 // target=native
 //
-// :6:11: error: type '@typeInfo(@typeInfo(@TypeOf(empty.b)).Fn.return_type.?).ErrorUnion.error_set!u32' is not indexable and not a range
+// :6:11: error: type '@typeInfo(@typeInfo(@TypeOf(tmp.b)).Fn.return_type.?).ErrorUnion.error_set!u32' is not indexable and not a range
 // :6:11: note: for loop operand must be a range, array, slice, tuple, or vector
 // :6:11: note: consider using 'try', 'catch', or 'if'

--- a/test/cases/compile_errors/for_loop_error_union.zig
+++ b/test/cases/compile_errors/for_loop_error_union.zig
@@ -1,0 +1,15 @@
+fn b() !u32 {
+    return 2;
+}
+
+export fn a() void {
+    for (b()) |_| {}
+}
+
+// error
+// backend=stage2
+// target=native
+//
+// :6:11: error: type '@typeInfo(@typeInfo(@TypeOf(empty.b)).Fn.return_type.?).ErrorUnion.error_set!u32' is not indexable and not a range
+// :6:11: note: for loop operand must be a range, array, slice, tuple, or vector
+// :6:11: note: consider using 'try', 'catch', or 'if'

